File: apps/sepuling-java/src/main/java/smecalculus/bezmen/storage/springdata/SepulkaRepository.java
package smecalculus.bezmen.storage.springdata;

import java.util.Optional;
import java.util.UUID;
import lombok.NonNull;
import org.springframework.data.jdbc.repository.query.Modifying;
import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import smecalculus.bezmen.storage.StateEm;

public interface SepulkaRepository extends CrudRepository<StateEm.AggregateRoot, String> {

    Optional<StateEm.Existence> findByExternalId(@NonNull String externalId);

    Optional<StateEm.Preview> findByInternalId(@NonNull UUID internalId);

    @Modifying
    @Query(
            """
            UPDATE sepulkas
            SET revision = revision + 1,
                updated_at = :#{#state.updatedAt}
            WHERE internal_id = :id
            AND revision = :#{#state.revision}
            """)
    int updateBy(@Param("state") StateEm.Touch state, @Param("id") UUID internalId);
}

File: apps/sepuling-kotlin/src/main/kotlin/smecalculus/bezmen/storage/springdata/SepulkaRepository.kt
package smecalculus.bezmen.storage.springdata

import org.springframework.data.jdbc.repository.query.Modifying
import org.springframework.data.jdbc.repository.query.Query
import org.springframework.data.repository.CrudRepository
import org.springframework.data.repository.query.Param
import smecalculus.bezmen.storage.StateEm
import java.util.UUID

interface SepulkaRepository : CrudRepository<StateEm.AggregateRoot, String> {
    fun findByExternalId(externalId: String): StateEm.Existence?

    fun findByInternalId(internalId: UUID): StateEm.Preview?

    @Modifying
    @Query(
        """
        UPDATE sepulkas
        SET revision = revision + 1,
            updated_at = :#{#state.updatedAt}
        WHERE internal_id = :id
        AND revision = :#{#state.revision}
        """,
    )
    fun updateBy(
        @Param("state") state: StateEm.Touch,
        @Param("id") internalId: UUID,
    ): Int
}
